llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -590,26 +588,35 @@ Value *LibCallSimplifier::optimizeStrCmp(CallInst *CI, IRBuilderBase &B) {
   if (Len1 && Len2) {
     return copyFlags(
         *CI, emitMemCmp(Str1P, Str2P,
-                        ConstantInt::get(DL.getIntPtrType(CI->getContext()),
-                                         std::min(Len1, Len2)),
+                        ConstantInt::get(
+                            DL.getIntPtrType(
+                                CI->getContext(),
+                                Str1P->getType()->getPointerAddressSpace()),
+                            std::min(Len1, Len2)),
```

Comment:
the memcmp case should probably have a similar wrapper

Reply:
If you're going to change this, please change it to use the correct value -- which is TLI::getSizeTSize().

Reply:
Unfortunately, `TLI::getSizeTSize()` contains a nice comment explaining that it always uses `addrspace(0)`, and noting that maybe it should consider alternatives, which is exactly the goal here.

Reply:
There's a few alternatives in here, none of them great:

* We can use the current state of this patch, which will have every target always use i64. As efriedma points out, this will probably work out because of backend legalization.
* We can thread TLI into `IRBuilder::CreateMemCpy` so that it can pick the right size. This requires updating a number of other passes throughout the tree, many of which don't depends on TargetLibraryInfo today.
* Same as above, but we add it in parallel to the existing overloads so that we can migrate users gradually.
* Build the types in SimplifyLibCalls as in the original version of this diff. Do some local cleanup to make it more readable per arsenm's feedback.

Thoughts?

Reply:
The reason getSizeTSize() unconditionally queries addrspace(0) is just that there aren't any in-tree targets where it needs to return anything else. Fixes welcome.

Passing in a integer to emitMemCmp where the size isn't getSizeTSize() just miscompiles, as far as I know.  (CreateMemCpy works fine in any case because it's an intrinsic.)

Reply:
> The reason getSizeTSize() unconditionally queries addrspace(0) is just that there aren't any in-tree targets where it needs to return anything else. Fixes welcome.

Honestly, I don't think fixing `getSizeTSize()` makes sense. It's just a wrapper around `DataLayout::getPointerSizeInBits`, which is already more widely accessible throughout the middle-end than `TargetLibraryInfo` is.

If we really want the convenience name, then I think we should move `getSizeTSize` to `DataLayout`, and add an addrspace argument such that it is a pure wrapper around `getPointerSizeInBits`.

Reply:
getSizeTSize() is supposed to be the bitwidth of the type size_t.  There's only one size_t.  This is useful because we're talking about C library functions, which are defined in terms of size_t.

If you want to query information about a specific address-space, we have appropriate datalayout APIs.

Reply:
Would you be amenable to moving `getSizeTSize()` onto `DataLayout`? That would make it much easier to update all callers of `CreateMemCpy`.

I'll look at ensuring `getSizeTSize()` is correct for CHERI targets in a separate PR.

Reply:
I don't think getSizeTSize belongs on the data layout. It's a property related to the C library, rather than the target itself.

For CHERI's purposes, shouldn't it be sufficient to consistently make use of getSizeTSize() in this file and patch getSizeTSize() to return the correct value? Maybe just making the default value be the index type size instead of the pointer type size would be sufficient?

The BuildLibCalls code itself already uses the correct types, we're just not passing in matching types here.

Reply:
> For CHERI's purposes, shouldn't it be sufficient to consistently make use of getSizeTSize() in this file and patch getSizeTSize() to return the correct value? Maybe just making the default value be the index type size instead of the pointer type size would be sufficient?

I've split such a change into https://github.com/llvm/llvm-project/pull/118747
There seems to be at least one test that concretely tests for the opposite behavior today, so we'll need to sort that out. If we can make that change, then most of the changes in this PR become irrelevant.